File: packages/mesh-io/typescript/src/read-mesh.ts
import {
  BinaryFile,
  Mesh,
  getFileExtension,
} from 'itk-wasm'

import mimeToMeshIo from './mime-to-mesh-io.js'
import extensionToMeshIo from './extension-to-mesh-io.js'
import meshIoIndex from './mesh-io-index.js'

import ReadMeshOptions from './read-mesh-options.js'
import ReadMeshResult from './read-mesh-result.js'

interface ReaderResult {
  webWorker: Worker,
  couldRead: boolean,
  mesh: Mesh
}
interface ReaderOptions {
  /** Only read mesh metadata -- do not read pixel data. */
  informationOnly?: boolean
}
type Reader = (webWorker: null | Worker | boolean, serializedMesh: File | BinaryFile, options: ReaderOptions) => Promise<ReaderResult>

/**
 * Read a mesh file format and convert it to the itk-wasm file format
 *
 * @param {webWorker} null | webWorker - Web worker to run the pipeline or null to run it in a new worker
 * @param {File | BinaryFile} serializedMesh - Input mesh serialized in the file format
 * @param {ReadMeshOptions} options - options to cast the resulting mesh type or to only read mesh metadata
 *
 * @returns {Promise<ReadMeshResult>} - result object with the mesh and the web worker used
 */
async function readMesh(
  webWorker: null | Worker | boolean,
  serializedMesh: File | BinaryFile,
  options: ReadMeshOptions = {}
) : Promise<ReadMeshResult> {

  const mimeType = (serializedMesh as File).type ?? ''
  const fileName = (serializedMesh as File).name ?? (serializedMesh as BinaryFile).path ?? 'fileName'
  const extension = getFileExtension(fileName).toLowerCase()
  let usedWebWorker = webWorker

  let serializedMeshFile = serializedMesh as BinaryFile
  if (serializedMesh instanceof Blob) {
    const serializedMeshBuffer = await serializedMesh.arrayBuffer()
    serializedMeshFile = { path: serializedMesh.name, data: new Uint8Array(serializedMeshBuffer) }
  }

  let io = null
  if (mimeType && mimeToMeshIo.has(mimeType)) {
    io = mimeToMeshIo.get(mimeType)
  } else if (extensionToMeshIo.has(extension)) {
    io = extensionToMeshIo.get(extension)
  } else {
    for (const readerWriter of meshIoIndex.values()) {
      if (readerWriter[0] !== null) {
        let { webWorker: testWebWorker, couldRead, mesh } = await (readerWriter[0] as unknown as Reader)(usedWebWorker, { path: serializedMeshFile.path, data: serializedMeshFile.data.slice() }, { informationOnly: options.informationOnly })
        usedWebWorker = testWebWorker
        if (couldRead) {
          return { webWorker: usedWebWorker, mesh }
        }
      }
    }
  }
  if (!io) {
    throw Error('Could not find IO for: ' + fileName)
  }
  const readerWriter = meshIoIndex.get(io as string)

  const reader = (readerWriter as Array<Reader>)[0]
  let { webWorker: testWebWorker, couldRead, mesh } = await reader(usedWebWorker, serializedMeshFile, { informationOnly: options.informationOnly })
  usedWebWorker = testWebWorker
  if (!couldRead) {
    throw Error('Could not read: ' + fileName)
  }

  return { webWorker: usedWebWorker, mesh }
}

export default readMesh

File: packages/mesh-io/python/itkwasm-mesh-io-wasi/itkwasm_mesh_io_wasi/read_mesh.py
import os
import importlib
from pathlib import Path

from itkwasm import Mesh

from .extension_to_mesh_io import extension_to_mesh_io
from .mesh_io_index import mesh_io_index

def read_mesh(
    serialized_mesh: os.PathLike,
    information_only: bool = False,
) -> Mesh:
    """Read a mesh file format and convert it to the ITK-Wasm file format.

    :param serialized_mesh: Input mesh serialized in the file format
    :type  serialized_mesh: os.PathLike

    :param information_only: Only read mesh metadata -- do not read pixel data.
    :type  information_only: bool

    :return: Output mesh
    :rtype:  Mesh
    """
    extension = ''.join(Path(serialized_mesh).suffixes)

    io = None
    if extension in extension_to_mesh_io:
        func = f"{extension_to_mesh_io[extension]}_read_mesh"
        mod_name = f"itkwasm_mesh_io_wasi.{func}"
        mod = importlib.import_module(mod_name)
        io = getattr(mod, func)
    else:
        for ioname in mesh_io_index:
            func = f"{ioname}_read_mesh"
            mod_name = f"itkwasm_mesh_io_wasi.{func}"
            mod = importlib.import_module(mod_name)
            io = getattr(mod, func)
            could_read, mesh = io(serialized_mesh, information_only=information_only)
            if could_read:
                return mesh

    if io is None:
        raise RuntimeError(f"Could not find an mesh reader for {extension}")

    could_read, mesh = io(serialized_mesh, information_only=information_only)
    if not could_read:
        raise RuntimeError(f"Could not read {serialized_mesh}")

    return mesh


def meshread(
    serialized_mesh: os.PathLike,
    information_only: bool = False,
) -> Mesh:
    return read_mesh(serialized_mesh, information_only=information_only)

meshread.__doc__ = f"""{read_mesh.__doc__}
    Alias for read_mesh.
    """